File: src/nylon/parser/parsers/BuiltinParser.java
package nylon.parser.parsers;

import nylon.builtins.objects.BuiltinFunction;
import nylon.parser.Parser;

/**
 * Created by Aedan Smith.
 */

public class BuiltinParser {
    public static Parser getParser(BuiltinFunction builtinFunction) {
        return (stringIterator, nylonParser) -> {
            stringIterator.skip();
            return builtinFunction;
        };
    }
}

File: src/nylon/parser/parsers/BuiltinParser.kt
package nylon.parser.parsers

import nylon.NylonFunction
import nylon.parser.CharIterator
import nylon.parser.NylonParser
import nylon.parser.Parser
import nylon.parser.ParserBuilder
import java.util.*
import java.util.function.Supplier

/**
 * Created by Aedan Smith.
 */

open class BuiltinParserBuilder(val nylonFunction: NylonFunction, val identifier: Char) : ParserBuilder {
    override fun accept(parsers: ArrayList<Parser>) {
        parsers[identifier.toInt()] = BuiltinParser(nylonFunction)
    }
}

class BuiltinParser(val nylonFunction: NylonFunction) : Parser {
    override fun apply(src: CharIterator, parser: NylonParser): NylonFunction? {
        src.next()
        val function = nylonFunction.clone()
        function.resolveArgs(Supplier { parser.parse(src)!! })
        return function
    }
}
